azerothcore/azerothcore-wotlk
# CreatureUnitRelocationWorker reworked

This is further support of void MoveInLineOfSight(Unit* who) override of creature AIs.
Currently before this PR, src scripted creatures of void MoveInLineOfSight(Unit* who) override would only follow void MoveInLineOfSight(Unit* who) override if they were hostile and nothing else. With this PR we create it with any aggression stance\faction stance, which will open up while improving upon only the future scripting of the core.



## Changes Proposed:
Ellminates restricting HasReactState(REACT_AGGRESSIVE) only to requirements of creatures to use void MoveInLineOfSight(Unit* who) override to having no React State requirements.

## Issues Addressed:
Issue was first addressed in discord message and PR of @malcrom 
https://github.com/azerothcore/azerothcore-wotlk/blob/58cfa8cf53dafdea2cba396a3f003e6e4df21d66/src/server/scripts/Northrend/zone_borean_tundra.cpp

https://discord.com/channels/217589275766685707/217589275766685707/900788241718976602
While running his PR in debugger mode i discovered that the void MoveInLineOfSight(Unit* who) override was not being hit at all. Later it was pointed out by @Nyeriah that the MoveInLineOfSight was only applying to creatures of aggress states (2), and not of defensive (1) or passive (0). 

A issue present only with azerothcore and not the TC counter part.



## SOURCE:
https://github.com/TrinityCore/TrinityCore/blob/e615b71509c018cb856680698f28bcdf83cde032/src/server/game/Grids/Notifiers/GridNotifiers.cpp#L190

## Tests Performed:
Builds with out errors.

Parts of this refactor is from TC which is tested and working with the pr https://github.com/azerothcore/azerothcore-wotlk/blob/58cfa8cf53dafdea2cba396a3f003e6e4df21d66/src/server/scripts/Northrend/zone_borean_tundra.cpp where at line 636 of the void update prisoner was changed to the three states of REACT_PASSIVE, REACT_DEFENSIVE , REACT_AGGRESSIVE. The desire outcome was finally the npc was despawn regardless of their reaction state.


## How to Test the Changes:
Test any Scripted Creature AI that has  void MoveInLineOfSight(Unit* who) override.
The PR of malcroms was the only one of its kinda on AZerothcore to have a script ai of none aggressive stance.
Feel free to adjust the the react state of other Source Scripted AI's void MoveInLineOfSight(Unit* who) with passive and defensive and see that Debugger will skip over the  void MoveInLineOfSight(Unit* who)  before this PR is applied. Applying this pr you will see that debugger will hit the  void MoveInLineOfSight(Unit* who)  and script will continue on as intended.

## Known Issues and TODO List:
- Futher independent testing. Although I firmly believe, this is a improvement only to the core for furture source scripting of AI's to use the void MoveInLineOfSight(Unit* who), due to its nature we need further testing to ensure regression would not be introduced as this will effect the core.


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
